### ISSUE TITLE
Discussion: REST vs GraphQL

### ISSUE THREAD

**Issue by QuiCM** [Pryaxis/TShock#1546]
See the associated project.

This issue is for discussion.
Please provide feedback, commentary, examples, etc. on how you'd like the REST system to be improved

**Comment by QuiCM** [comment]
Copied from #1543 (Some REST endpoints should be used with POST instead of GET) for consolidation:

>Ryozuki:
The endpoints that in my opinion should be using POST instead of GET:

    /bans/create
    /v2/server/broadcast
    /v2/server/off
    /v3/server/rawcmd
    /v3/server/reload
    /v3/server/restart
    /v2/users/create
    /v2/users/destroy
    /v2/users/update

**Comment by hakusaro** [comment]
All /v2/ endpoints need to be removed or upgraded to /v3/.

Long term, remove REST and replace it with [GraphQL](http://graphql.org). It's kind of taken off as the standard replacement for RESTful API development. This is mostly because it enables practically infinite types of queries without writing new routes for every type of query.

**Comment by QuiCM** [comment]
Having briefly looked at GraphQL, LGTM

**Comment by Ryozuki** [comment]
I'll look at GraphQL, i would like to try to do this rewrite.

**Comment by hakusaro** [comment]
Blocking this on fragments & cleanup of internal code.

**Comment by XGhozt00** [comment]
I'm for replacing REST. I only ask that we deprecate the REST API endpoints _before_ removing them completely so I have a chance to update things without an outage. Ideally just leave it in there for a few releases with a EOL scheduled date. Also, please document any changes to the REST API endpoints in the release notes (such as changes GET to POST).

This will be an awesome update! Thanks for working so hard on this.

**Comment by Ryozuki** [comment]
Sure

**Comment by Ijwu** [comment]
GraphQL isn't necessarily a replacement for REST, it's an alternative. Let's keep this in mind before jumping on the wagon. GraphQL is a query language first and foremost. We expose data in our REST API but there's also a lot of action endpoints.

GraphQL implements something called [Mutations](http://graphql.org/learn/queries/#mutations) but those don't necessarily cover all of our use cases. For example, the [/v3/server/rawcmd](https://tshock.readme.io/v4.3.24/reference#v3serverrawcmd) endpoint is not necessarily a data transformation and its output varies by command usage. This isn't necessarily complementary to GraphQL usage.

https://philsturgeon.uk/api/2017/01/24/graphql-vs-rest-overview/

**Comment by XGhozt00** [comment]
That rawcmd endpoint is very important to me.

**Comment by hakusaro** [comment]
Iirc you can write mutations any way you like. You could just have a
mutation that emulates it.

**Comment by Ryozuki** [comment]
I actually like more REST than GraphQL

**Comment by hakusaro** [comment]
Great.

**Comment by hakusaro** [comment]
I prefer Coke to Pepsi.

**Comment by Ryozuki** [comment]
Hey me too!

**Comment by Ryozuki** [comment]
With a REST API, if I retrieve a user’s data with GET /users/{id}, I can try to update it with PATCH /users/{id} and delete it with DELETE /users/{id} without even reading the documentation and it will probably work… like it will work with another another resource in the same API and even with another resource in another REST API. And icing on the cake: both humans and machines can discover and understand the meaning of each operation. Using the HTTP protocol enforce a certain consistency and allow predictability.

In a GraphQL API, reading and writing actions are separated in 2 sets of queries.

    query for reading
    mutation for writing

Updating or deleting this user with GraphQL will not be as clear as with REST. How will be named the mutation allowing me to delete a user? deleteUser, removeUser or suppressPeople? You’ll have to dig in the documentation to find out. You cannot guess it because different API providers will have different naming conventions and these naming convention may event not be consistent within an API.

Being protocol agnostic also means that everything is going to be 200 OK when using GraphQL over HTTP, wheither the query was OK or not. With REST APIs we are acustomed to be able to tell what happens just by looking at the HTTP status. Whatever the API, if we receive a 404 HTTP status we known what it means: resource not found.

 source: https://apihandyman.io/and-graphql-for-all-a-few-things-to-think-about-before-blindly-dumping-rest-for-graphql/

**Comment by hakusaro** [comment]
>  I can try to update it with PATCH /users/{id} and delete it with DELETE /users/{id} without even reading the documentation and it will probably work… like it will work with another another resource in the same API and even with another resource in another REST API

Currently, 100% of the REST API _in TShock_ is non-conformant with this idea anyway. All of the queries have to be GETs. Commands that already mutate the server state are already destructive on GET.

> Updating or deleting this user with GraphQL will not be as clear as with REST. How will be named the mutation allowing me to delete a user? deleteUser, removeUser or suppressPeople? You’ll have to dig in the documentation to find out. You cannot guess it because different API providers will have different naming conventions and these naming convention may event not be consistent within an API.

You already can't guess with our REST implementation. None of the routes are named standard things, and the resources returned aren't RESTful at all. If you want to `DELETE` a group you have to call `/v2/groups/destroy/` with your group (on a `GET`). If you want to ban a player (a destructive action), then you have to call `/v2/players/ban/` with `GET` and (off the top of my head) I don't know what to even send it. With groups, I knew I could send a group name by inference. What about player? You can ban people by IP but how does that work in this API?

REST verbs are great when your goal is to modify all user state info at once and map everything, but since we don't do that, you have to rely on us for things like `/world/meteor` which is already arbitrary, non-conformant with REST, and non-trivial to model correctly in either the REST or the GraphQL world.

There's no verb in HTTP for `mute player` so we have a `GET` on `/v2/players/mute` instead.

**Comment by XGhozt00** [comment]
I agree with @Ryozuki -- in typical REST API environments, that's how it should work and it is easy to follow because most endpoints are handled with POST and so you don't have to lookup all the endpoints because the POST data can change what you want to do and the rest of it is usually obvious. That also includes PATCH and DELETE methods.

The problem with this is..... tshock's REST API in it's current state doesn't work like this at all. It's a huge mess, trust me.. I'v used every single endpoint and found all the bugs and inconsistencies. That includes the endpoints, expected data received, etc..

Should they stick with REST, /v4/ might as well be a complete overhaul. And if you're overhauling it, you might as well research alternative solutions. Enter GraphQL.

When you then consider that tshock is built by volunteers <3, I typically lean towards what's easier for them to create and manage, as that encourages development and keeps the software alive. No one wants to do grunt work for free.

**Comment by hakusaro** [comment]
@XGhozt00 to be clear: as the largest or close to largest REST user, you have pretty much ultimate say on what goes and doesn't go with any system rewrite (we want it to be easy on you to match too). We ultimately want to move to GraphQL for the ability to provide longer term access to more data types than we currently support (limited basic management functionality). You shouldn't have to depend on rawcmd for anything other than a user wanting to run raw commands, ideally. There should be a systematic way of querying things sanely that you can use in 99% of cases.

**Comment by XGhozt00** [comment]
I really appreciate that @hakusaro -- however, I don't have enough experience with GraphQL to vouch for it. It looks a little complicated but I'm sure I could figure it out once we have something to play with. I will try to catch up on some reading about interfacing with it when I get a chance this week/end.

I haven't looked, but... has anyone considered maybe just creating a new branch here and testing it out? Even if only one thing works.. like getting server status or player list. Even if the codes a mess, maybe we love it, maybe we hate it? I don't imagine that would take too long to do?

**Comment by bartico6** [comment]
I'll just hop in and ask the obvious question - what is the exact reason you want to replace REST with GraphQL for, other than "exciting new tech"? I don't really see a reason to replace a (perfectly usable) standard in tShock for a new standard that:  
1) almost nobody of the end users is aware of or decent with  
2) doesn't even seem intuitive for things like, say, adding bans or running console commands  
3) at first sight seems really ambiguous in terms of failures (what do we even expect?)  
  
The fact the REST code is bad (in my opinion) is just incentive to rewrite it - to replace it entirely with something completely different due to bad code is like rolling out a cannon to kill a mosquito.  
Don't take offense to this but (since nobody really argued a case for GraphQL so far) it really seems like an unnecessary change that might have potential in the future but so far it offers us nothing new or useful as a concept and even throws hurdles our way...

**Comment by Ijwu** [comment]
This PR discusses, essentially, starting fresh. Clean slate rebuild. 

It doesn't matter what tech we use as long as we understand the implications and can build it out reasonably. If GraphQL is easier to build, easier to consume, and covers all use cases: cool. 

If someone can prove to me SOAP can do that, I'd even aim for that. 

I don't see a reason to stick to REST other than 'it's already here'. You're going to have to do better than shoot the shiny new tech just for being so.

**Comment by hakusaro** [comment]
At this point, since the rest plugin is so bad, I think there are two options:

1. Remove REST from TShock and provide a drop-in plugin that will provide the current REST implementation as is (same functionality, but out of this codebase so it's easier to maintain).
2. Remove REST and move to GraphQL (a solution I offer to do).

Everyone who has contributed to this discussion thus far has not volunteered for the task of doing _any of the proposed work_ that I intended to have completed prior to the beginning of February 2018.

If people want to step up to volunteer to write a standards compliant rest implementation be my guest. Fragments might not even have to be merged for that to happen. That means it has to fully replace, provide backwards compatibility, etc. for the current backend in a sane way. In short, it has to do what my proposed GraphQL implementation would do (provide a bridge timeframe for people to switch to the new system and make sure the system is 100% compatible in functionality to the existing system).

I, however, wrote the initial REST implementation, and do not want it around anymore. It was a mistake for me to have implemented it without first considering the time required to maintain it as well as the technical skill to provide a useful backend for consumption in a way consistent with good REST design.

Right now, the person who has signed onboard to do this rewrite is me. My solution was GraphQL. I offered to do the implementation work to get it off the ground and to get us from A->B. Because it's a more sane solution for me to deal with. If I'm maintaining it, then I want to get something practical out of it (experience in software design) -- not just become the manager of the REST rewrite.

Unless you are a consumer of the REST API (in which case, you're basically only @XGhozt00), giving input on what we switch to isn't really useful. Setting quantifiable targets is. The people who have a right to say anything about this are people who have consumed it in the past and people who have to maintain it.

It's really easy to bikeshed this problem. Let's avoid continuing down that path. If you are proposing a solution that _you_ want to take responsibility over, please make that clear. If you are a concerned user, we can move timelines and make sure that everything is cozy for the eventual removal of the REST API. If you are neither, however, throwing sand at GraphQL for the sake of throwing sand is stupid.

> what is the exact reason you want to replace REST with GraphQL for, other than "exciting new tech"?

We've already covered this. It appears to provide a more transparent way to tag our data structures as accessible over the network for querying than trying to decide what users want to query and writing more REST routes to maintain them.

Also, it would look good on maintainers' resumes that they implemented GraphQL, a solution that the industry is moving to (don't believe me? See Facebook and GitHub).

> The fact the REST code is bad (in my opinion) is just incentive to rewrite it - to replace it entirely with something completely different due to bad code is like rolling out a cannon to kill a mosquito.

To rewrite REST would require nearly all of the code changing that handles REST, and on top of that, would require re-implementation and code maintenance and backwards compatibility with the old system to phase people in that borders on insanity. When it comes time for a project to consider rewriting a codebase the move should consider possibilities that show promise (like GQL).

> Don't take offense to this but (since nobody really argued a case for GraphQL so far that was convincing more than "expose more stuff") it really seems like an unnecessary change that might (and doesn't necessarily have to) have potential in the future but so far it offers us nothing new or useful as a concept and even throws hurdles our way...

You aren't offering to implement a REST rewrite. I am offering to implement GraphQL and phase out REST in a sane way. If you think this won't happen, I urge you to look at the lengths we've gone to in the past for migrating users to new systems and making effective decisions about deprecation and removals. In particular, we kept deprecated REST routes around for months longer than we promised for the sake of ensuring people were 100% fully switched.

> I haven't looked, but... has anyone considered maybe just creating a new branch here and testing it out? Even if only one thing works.. like getting server status or player list. Even if the codes a mess, maybe we love it, maybe we hate it? I don't imagine that would take too long to do?

Like fragments, GraphQL would start its life in a branch, eventually gaining features to have parity with REST. At no time would REST be "replaced" without achieving that level of parity. This issue is reflective of the project's end goal, not reflective of the steps in between. There's a lot of work that goes into TShock prior to each release, and it's not as if we'd ship a new API with 30% of the features of the old one.

_tl;dr: Users can provide feedback on migration and ease of use of GQL after experimenting with it. Developers can continue to discuss efforts to move to GQL and what data we want to expose. Developers offering to volunteer to do the rewrite (like I am) can volunteer to do a full REST system rewrite. If this happens, we will rebuild conversations to focus on what needs to be redone in the new REST system._

**Comment by hakusaro** [comment]
More on the "we need a backwards compatibility layer" front: Multiplay probably still uses REST.

**Comment by hakusaro** [comment]
Gregor confirmed MPUK still uses REST.

**Comment by hakusaro** [comment]
Somewhat waiting on Enerdy to confirm if he wants to take this issue. He claims he'll volunteer on the 22nd, but I have no idea.

**Comment by Enerdy** [comment]
So this is what I'll be doing:

1. Make a `strip-rest` branch where I strip REST out of TShock.
2. Make a new repository for a RESTAPI TShock Plugin, where the current version of REST will be re-implemented to maintain backwards compatibility.
3. Merge the `strip-rest` branch, and start shipping the RESTAPI Plugin with TShock OR make it clear in the Release Notes that REST users will have to install a new plugin if they wish to continue to use REST.
4. Start implementing all of our current REST features in the RESTAPI Plugin in a more standard compliant way (not everything is a GET), and eventually deprecate routes v1-3 so that they may be removed after giving people enough time to adapt.

The goal is to remove the drag of keeping REST up to date from the TShock repository, and instead make it an independent project as a plugin. GraphQL will still be implemented by @hakusaro in some shape or form. This gives users the ability to have both systems active, or just one of them, or opt-out from GraphQL's configuration and just have neither.

**Comment by QuiCM** [comment]
Tracking for REST plugin is now in #1588.
Closing this
